_posts/2023-03-20-cyclonedx-maven-aggregate-bom-why-not-to-trust.md

```diff
@@ -0,0 +1,376 @@
+
+```
+<component type="library"
+    bom-ref="pkg:maven/org.owasp.webgoat.lesson/auth-bypass@v8.0.0.M15?hash=<hash1>&amp;type=jar">
+  <publisher>OWASP</publisher>
+  <group>org.owasp.webgoat.lesson</group>
+  <name>auth-bypass</name>
+  <version>v8.0.0.M15</version>
+  <description>Parent Pom for the WebGoat Project. A deliberately insecure Web Application</description>
+  <purl>pkg:maven/org.owasp.webgoat.lesson/auth-bypass@v8.0.0.M15?type=jar</purl>
+  ...
+</component>
+
+<component type="library"
+    bom-ref="pkg:maven/org.owasp.webgoat.lesson/auth-bypass@v8.0.0.M15?hash=<hash2>&amp;type=jar">
+  <publisher>OWASP</publisher>
+  <group>org.owasp.webgoat.lesson</group>
+  <name>auth-bypass</name>
+  <version>v8.0.0.M15</version>
+  <description>Parent Pom for the WebGoat Project. A deliberately insecure Web Application</description>
+  <purl>pkg:maven/org.owasp.webgoat.lesson/auth-bypass@v8.0.0.M15?type=jar</purl>
+  ...
+</component>
+```
+
+It is this change which, while not invalid according to the specification schema, will likely cause problems for tools which have made the assumption the component can only exist once.
+
+# Possible Future Direction in the Specification
+
+While trying to work out the best approach for solving this problem, Steve Springett, Hervé Boutemy and I spent time on a video call discussing whether the [CycloneDX maven plugin PR](https://github.com/CycloneDX/cyclonedx-maven-plugin/pull/306) I had submitted was the right approach and, if not, could we think of other possible solutions.  Part of this discussion involved deficiencies in the specification and how the specification could be updated to allow support for alternate dependency hierarchies.
+
+The problem within the specification really comes down to the fact the references used within the `dependencies` section are component references, and as such set up a 1-1 relationship between the dependency hierarchy and the component.  For example,
+
+```
+<components>
+  <component type="library" bom-ref="component1-bom-ref">
+  ...
+  </component>
+  <component type="library" bom-ref="component2-bom-ref">
+  ...
+  </component>
+</components>
+<dependencies>
+  <dependency ref="component1-bom-ref">
+    <dependency ref="component2-bom-ref"/>
+  </dependency>
+  <dependency ref="component2-bom-ref"/>
+</dependencies>
+```
+The above structure does not allow for a second top level dependency element to reference the same component since the `ref` attribute is the key to retrieve the dependency information.
+
+A solution for this would be a change to the specification to define references in the dependency section to be uniquely scoped to within that section, distinct from those used within the `component` section, and require a second reference to the `component`.  For example,
```

Comment:
The JSON Schema is different and this approach will not work. Not sure about protobuf.